GH-3107: Add errorOnTimeout for Inbound Gateways

artembilan opened the pull request:
Fixes https://github.com/spring-projects/spring-integration/issues/3107

The `MessagingGatewaySupport` has an `errorOnTimeout` option to throw
a `MessageTimeoutException` when downstream reply doesn't come back in
time for configured reply timeout

* Expose an `errorOnTimeout` option as a `TcpInboundGateway` ctor
property
* Add new factory methods into a `Tcp` factory for Java DSL
* Ensure a property works as expected in the `IpIntegrationTests`
* Document a new option

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-3107: Add errorOnTimeout for TcpInboundGateway

Fixes https://github.com/spring-projects/spring-integration/issues/3107

The `MessagingGatewaySupport` has an `errorOnTimeout` option to throw
a `MessageTimeoutException` when downstream reply doesn't come back in
time for configured reply timeout

* Expose an `errorOnTimeout` option as a `TcpInboundGateway` ctor
property
* Add new factory methods into a `Tcp` factory for Java DSL
* Ensure a property works as expected in the `IpIntegrationTests`
* Document a new option

artembilan commented:
I wonder if we should just go ahead and do the same for all the Inbound Gateway implementations...

garyrussell commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpInboundGateway.java:
```diff
@@ -117,7 +139,7 @@ public boolean onMessage(Message<?> message) {
 
 	private boolean doOnMessage(Message<?> message) {
-		Message<?> reply = this.sendAndReceiveMessage(message);
+		Message<?> reply = sendAndReceiveMessage(message);
```
Need to catch and swallow `MessageTimeoutException` to retain existing behavior when no EC or error flow returns `null`.

artembilan replied:
Well, what we have here is fully a new behavior which doesn't break an existing one by default because all the existing `TcpInboundGateways` are created without that new `errorOnTimeout` ctor.
Also see an explanation of that option in the super class:
```
  If no error channel is
	 * configured, the {@link MessageTimeoutException} will be thrown.
```

So, what you are asking is fully redundant since it is related only to new option and we definitely should throw that `MessageTimeoutException` according contact. And again: we do that only if `errorOnTimeout` is turned on, which is not by default.

Does it make sense?

garyrussell replied:
Good point; I was concerned about generating ERROR logs in the caller to `onMessage` but, as you say, they have to opt-in by setting the boolean, in which case an error log is appropriate anyway.

garyrussell reviewed (changes requested):
Yes, we should probably do it for all in gates - this PR? Or open another "backlog" issue?

artembilan commented:
> do it for all in gates - this PR

Yes, I can play with that in this PR: doesn't look like we are in hurry to release this fix.
Although I'd follow your review preferences.
If that easier for you to merge this and go ahead with other in other ticket, I can do that as well.

garyrussell commented:
I don't mind if you add more here.

artembilan committed:
* Add a setter for MessagingGatewaySupport.errorOnTimeout option
* Expose an `errorOnTimeout` option on the DSL's `MessagingGatewaySpec`
making all the out-of-the-box inbound gateways possible to react to the
`MessageTimeoutException` when no reply during reply timeout
* Propagate properly `errorOnTimeout` in the `JmsInboundGateway`
* Modify docs respectively

artembilan commented:
OK. So, pushed a general solution to configure any inbound gateway for `errorOnTimeout`.

garyrussell commented on src/reference/asciidoc/endpoint-summary.adoc:
```diff
@@ -223,3 +223,6 @@ The `<int:outbound-channel-adapter>` element lets you send data to a `void` meth
 Each of these works without requiring any source-level dependencies on Spring Integration.
 The equivalent of an outbound gateway in this context is using a service activator (see <<./service-activator.adoc#service-activator,Service Activator>>) to invoke a method that returns an `Object` of some kind.
+
+Starting with version `5.2.2`, all the inbound gateway can be configured with an `errorOnTimeout` boolean flag to throw a `MessageTimeoutException` when downstream reply doesn't come back during reply timeout.
```
`s/gateway/gateways/`

...when **the** downstream flow doesn't return a reply during the reply timeout.
The timer is not started until the thread returns the gateway so usually is only useful when the downstream flow is asynchronous.

garyrussell reviewed (changes requested):
LGTM with doc polish. Let's see if Travis agrees 😄

artembilan committed:
* Improve docs about `errorOnTimeout`

garyrussell commented:
Merged as 72f7c723920d0e10f5d6ac38f4231b61eabe52ec with a bit more polishing.
